instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java

```diff
@@ -19,45 +17,50 @@
 import java.util.stream.Collectors;
 import java.util.stream.StreamSupport;
 
-public abstract class Instrumenter<REQUEST, RESPONSE> {
+public /* sealed */ class Instrumenter<REQUEST, RESPONSE> {
+
+  public static <REQUEST, RESPONSE> InstrumenterBuilder<REQUEST, RESPONSE> newBuilder(
+      OpenTelemetry openTelemetry, String instrumentationName) {
+    return new InstrumenterBuilder<>(openTelemetry, instrumentationName);
+  }
 
   private final Tracer tracer;
   private final SpanNameExtractor<? super REQUEST> spanNameExtractor;
+  private final SpanKind spanKind;
   private final StatusExtractor<? super REQUEST, ? super RESPONSE> statusExtractor;
   private final List<AttributesExtractor<? super REQUEST, ? super RESPONSE>> extractors;
+  private final ErrorCauseExtractor errorCauseExtractor;
 
-  protected Instrumenter(
+  Instrumenter(
       Tracer tracer,
       SpanNameExtractor<? super REQUEST> spanNameExtractor,
+      SpanKind spanKind,
       StatusExtractor<? super REQUEST, ? super RESPONSE> statusExtractor,
-      Iterable<? extends AttributesExtractor<? super REQUEST, ? super RESPONSE>> extractors) {
+      Iterable<? extends AttributesExtractor<? super REQUEST, ? super RESPONSE>> extractors,
+      ErrorCauseExtractor errorCauseExtractor) {
     this.tracer = tracer;
     this.spanNameExtractor = spanNameExtractor;
+    this.spanKind = spanKind;
     this.statusExtractor = statusExtractor;
     this.extractors =
         StreamSupport.stream(extractors.spliterator(), false).collect(Collectors.toList());
+    this.errorCauseExtractor = errorCauseExtractor;
   }
 
   public Context start(Context parentContext, REQUEST request) {
-    SpanKind kind = spanKind(request);
     SpanBuilder spanBuilder =
-        tracer.spanBuilder(spanName(request)).setSpanKind(kind).setParent(parentContext);
+        tracer
+            .spanBuilder(spanNameExtractor.extract(request))
+            .setSpanKind(spanKind)
+            .setParent(parentContext);
 
-    AttributesBuilder attributesBuilder = Attributes.builder();
     for (AttributesExtractor<? super REQUEST, ? super RESPONSE> extractor : extractors) {
-      extractor.onStart(attributesBuilder, request);
+      extractor.onStart(spanBuilder::setAttribute, request);
```

Comment:
I'd like to stick with `AttributesBuilder` since while we still have the tension between Labels and Attributes, I'm sure we're going to use a single class to accumulate the tags for both spans and meters.

Reply:
👍